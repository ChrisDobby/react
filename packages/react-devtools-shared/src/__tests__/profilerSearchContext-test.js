/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import typeof ReactTestRenderer from 'react-test-renderer';
import type {FrontendBridge} from 'react-devtools-shared/src/bridge';
import type {
  DispatcherContext,
  StateContext,
} from 'react-devtools-shared/src/devtools/views/Profiler/ProfilerSearchContext';
import type Store from 'react-devtools-shared/src/devtools/store';

describe('ProfilerSearchContext', () => {
  let React;
  let ReactDOM;
  let TestRenderer: ReactTestRenderer;
  let bridge: FrontendBridge;
  let store: Store;
  let utils;

  let BridgeContext;
  let ProfilerContextController;
  let StoreContext;
  let TreeContextController;
  let ProfilerSearchContext;

  let dispatch: DispatcherContext;
  let state: StateContext;

  beforeEach(() => {
    utils = require('./utils');
    utils.beforeEachProfiling();

    bridge = global.bridge;
    store = global.store;
    store.collapseNodesByDefault = false;
    store.recordChangeDescriptions = true;

    React = require('react');
    ReactDOM = require('react-dom');
    TestRenderer = utils.requireTestRenderer();

    BridgeContext = require('react-devtools-shared/src/devtools/views/context')
      .BridgeContext;
    ProfilerContextController = require('react-devtools-shared/src/devtools/views/Profiler/ProfilerContext')
      .ProfilerContextController;
    StoreContext = require('react-devtools-shared/src/devtools/views/context')
      .StoreContext;
    TreeContextController = require('react-devtools-shared/src/devtools/views/Components/TreeContext')
      .TreeContextController;
    ProfilerSearchContext = require('react-devtools-shared/src/devtools/views/Profiler/ProfilerSearchContext');
  });

  const Capture = () => {
    dispatch = React.useContext(
      ProfilerSearchContext.ProfilerSearchDispatcherContext,
    );
    state = React.useContext(ProfilerSearchContext.ProfilerSearchStateContext);
    return null;
  };

  const Contexts = ({
    defaultSelectedElementID = null,
    defaultSelectedElementIndex = null,
  }: any) => (
    <BridgeContext.Provider value={bridge}>
      <StoreContext.Provider value={store}>
        <TreeContextController
          defaultSelectedElementID={defaultSelectedElementID}
          defaultSelectedElementIndex={defaultSelectedElementIndex}>
          <ProfilerContextController>
            <ProfilerSearchContext.ProfilerSearchContextController>
              <Capture />
            </ProfilerSearchContext.ProfilerSearchContextController>
          </ProfilerContextController>
        </TreeContextController>
      </StoreContext.Provider>
    </BridgeContext.Provider>
  );

  const selectNode = () => {};

  it('should find elements in the chart rows', () => {
    const chartRows = [
      [{id: 1, name: 'Foo'}],
      [{id: 2, name: 'Bar'}, {id: 3, name: 'Baz'}],
    ];

    const Foo = () => null;
    const Bar = () => null;
    const Baz = () => null;

    utils.act(() =>
      ReactDOM.render(
        <React.Fragment>
          <Foo />
          <Bar />
          <Baz />
        </React.Fragment>,
        document.createElement('div'),
      ),
    );

    let renderer;
    utils.act(() => (renderer = TestRenderer.create(<Contexts />)));
    expect(state).toMatchSnapshot('1: initial state');

    utils.act(() =>
      dispatch({
        type: 'SET_CHART_DATA',
        payload: {data: chartRows, selectNode},
      }),
    );

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'ba'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('2: search for "ba" - multiple match');

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'f'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('3: search for "f" - single match');

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'y'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('4: search for "y" - no match');
  });

  it('should find elements in the chart nodes', () => {
    const chartNodes = [
      {id: 1, name: 'Foo'},
      {id: 2, name: 'Bar'},
      {id: 3, name: 'Baz'},
    ];

    const Foo = () => null;
    const Bar = () => null;
    const Baz = () => null;

    utils.act(() =>
      ReactDOM.render(
        <React.Fragment>
          <Foo />
          <Bar />
          <Baz />
        </React.Fragment>,
        document.createElement('div'),
      ),
    );

    let renderer;
    utils.act(() => (renderer = TestRenderer.create(<Contexts />)));
    expect(state).toMatchSnapshot('1: initial state');

    utils.act(() =>
      dispatch({
        type: 'SET_CHART_DATA',
        payload: {data: chartNodes, selectNode},
      }),
    );

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'ba'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('2: search for "ba" - multiple match');

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'f'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('3: search for "f" - single match');

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'y'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('4: search for "y" - no match');
  });

  it('should select the next and previous items within the search results', () => {
    const chartNodes = [
      {id: 1, name: 'Foo'},
      {id: 2, name: 'Bar'},
      {id: 3, name: 'Baz'},
      {id: 4, name: 'Bar'},
      {id: 5, name: 'Baz'},
    ];

    const Foo = () => null;
    const Bar = () => null;
    const Baz = () => null;

    utils.act(() =>
      ReactDOM.render(
        <React.Fragment>
          <Foo />
          <Bar />
          <Baz />
          <Bar />
          <Baz />
        </React.Fragment>,
        document.createElement('div'),
      ),
    );

    let renderer;
    utils.act(() => (renderer = TestRenderer.create(<Contexts />)));
    expect(state).toMatchSnapshot('1: initial state');

    utils.act(() =>
      dispatch({
        type: 'SET_CHART_DATA',
        payload: {data: chartNodes, selectNode},
      }),
    );

    utils.act(() => dispatch({type: 'SET_SEARCH_TEXT', payload: 'ba'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('2: search for "ba"');

    utils.act(() => dispatch({type: 'GO_TO_NEXT_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('3: go to second result');

    utils.act(() => dispatch({type: 'GO_TO_NEXT_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('4: go to third result');

    utils.act(() => dispatch({type: 'GO_TO_PREVIOUS_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('5: go to second result');

    utils.act(() => dispatch({type: 'GO_TO_PREVIOUS_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('6: go to first result');

    utils.act(() => dispatch({type: 'GO_TO_PREVIOUS_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('7: wrap to last result');

    utils.act(() => dispatch({type: 'GO_TO_NEXT_SEARCH_RESULT'}));
    utils.act(() => renderer.update(<Contexts />));
    expect(state).toMatchSnapshot('8: wrap to first result');

    store.getElementIDAtIndex(0);
  });
});
