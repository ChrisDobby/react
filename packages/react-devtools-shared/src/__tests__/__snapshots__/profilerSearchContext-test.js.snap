// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilerSearchContext should find elements in the chart nodes: 1: initial state 1`] = `
Object {
  "chartData": null,
  "searchIndex": null,
  "searchResults": Array [],
  "searchText": "",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart nodes: 2: search for "ba" - multiple match 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart nodes: 3: search for "f" - single match 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
  ],
  "searchText": "f",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart nodes: 4: search for "y" - no match 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
  ],
  "searchIndex": null,
  "searchResults": Array [],
  "searchText": "y",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart rows: 1: initial state 1`] = `
Object {
  "chartData": null,
  "searchIndex": null,
  "searchResults": Array [],
  "searchText": "",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart rows: 2: search for "ba" - multiple match 1`] = `
Object {
  "chartData": Array [
    Array [
      Object {
        "id": 1,
        "name": "Foo",
      },
    ],
    Array [
      Object {
        "id": 2,
        "name": "Bar",
      },
      Object {
        "id": 3,
        "name": "Baz",
      },
    ],
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart rows: 3: search for "f" - single match 1`] = `
Object {
  "chartData": Array [
    Array [
      Object {
        "id": 1,
        "name": "Foo",
      },
    ],
    Array [
      Object {
        "id": 2,
        "name": "Bar",
      },
      Object {
        "id": 3,
        "name": "Baz",
      },
    ],
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
  ],
  "searchText": "f",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should find elements in the chart rows: 4: search for "y" - no match 1`] = `
Object {
  "chartData": Array [
    Array [
      Object {
        "id": 1,
        "name": "Foo",
      },
    ],
    Array [
      Object {
        "id": 2,
        "name": "Bar",
      },
      Object {
        "id": 3,
        "name": "Baz",
      },
    ],
  ],
  "searchIndex": null,
  "searchResults": Array [],
  "searchText": "y",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 1: initial state 1`] = `
Object {
  "chartData": null,
  "searchIndex": null,
  "searchResults": Array [],
  "searchText": "",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 2: search for "ba" 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 3: go to second result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 1,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 4: go to third result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 2,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 5: go to second result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 1,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 6: go to first result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 7: wrap to last result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 3,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;

exports[`ProfilerSearchContext should select the next and previous items within the search results: 8: wrap to first result 1`] = `
Object {
  "chartData": Array [
    Object {
      "id": 1,
      "name": "Foo",
    },
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchIndex": 0,
  "searchResults": Array [
    Object {
      "id": 2,
      "name": "Bar",
    },
    Object {
      "id": 3,
      "name": "Baz",
    },
    Object {
      "id": 4,
      "name": "Bar",
    },
    Object {
      "id": 5,
      "name": "Baz",
    },
  ],
  "searchText": "ba",
  "selectNode": [Function],
}
`;
